### DISCUSSION TITLE
💭 Enhancing 🧠 Logic query response formats

### DISCUSSION THREAD

**Discussion by ccamel** [okp4/community#14] [OKP4 Protocol]
## Context

The `logic` module, a component of the OKP4 blockchain, enables the compilation of Prolog programs and the submission of queries on-chain. Defined by a protobuf contract, this module mandates a specific formalism for response delivery, enabling access to a series of substitutions that fulfill the given query.

We're currently evaluating the optimal structuring of `logic` module query responses. The primary challenge is to find a balance among intricacy, detail, and interoperability in these responses.

The objective of this discussion is to deliberate on potential options and assimilate community feedback to augment the response formats of the `logic` module. This is a precursor to the development of an Architectural Decision Record (ADR).

#### Overview

This approach involves returning substitution values as structured terms, each comprising a value and, possibly, arguments.

- `term`: Denotes a Prolog data element, which might be an atom, integer, float, or compound.
- `name`: Denotes the term's identifier.
- `arguments`: An array of term entities, depicting the term's arguments. This recursive architecture facilitates the representation of complex, nested terms.

**Example:**

```json
[
  {
    "substitutions": [
      {
        "variable": "X",
        "term": {
          "name": "1"
        }
      },
      {
        "variable": "Y",
        "term": {
          "name": "bar1"
        }
      }
    ]
  },
  {
    "substitutions": [
      {
        "variable": "X",
        "term": {
          "name": "2"
        }
      },
      {
        "variable": "Y",
        "term": {
          "name": "bar2"
        }
      }
    ]
  }
]
```

In this paradigm, terms are depicted as a tree, including lists. For instance, a list like X as [1,2,3] would be represented as:

```json
[
  {
    "substitutions": [
      {
        "variable": "X",
        "term": {
          "name": ".",
          "arguments": [
            {
              "name": "1"
            },
            {
              "name": ".",
              "arguments": [
                {
                  "name": "2"
                },
                {
                  "name": ".",
                  "arguments": [
                    {
                      "name": "3"
                    },
                    {
                      "name": "[]"
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
]
```

This format also accommodates strings, typically perceived as lists in Prolog, through a more condensed representation using quotation marks.

Example:

```json
[
  {
    "substitutions": [
      {
        "variable": "X",
        "term": {
          "name": "\"foo bar\""
        }
      }
    ]
  }
]
```

#### Consequences

**Positive**

- Offers comprehensive detail, allowing clients to reconstruct original Prolog terms without additional parsing.

**Negative**

- Tends to be verbose and potentially less readable for humans.
- Can lead to large payload sizes, especially with extensive lists.

**Neutral**

None

#### Overview

This format presents responses as strings, directly representing the substitution values as Prolog terms.

Example:

```json
[
  {
    "substitutions": [
      {
        "variable": "X",
        "expression": "1"
      },
      {
        "variable": "Y",
        "expression": "[1,2,3]"
      },
      {
        "variable": "Z",
        "expression": "\"foo bar\""
      }
    ]
  }
]
```

#### Consequences

**Positive**

- More compact, resulting in smaller payload sizes.
- Easier to comprehend for human users.

**Negative**

- Necessitates client-side parsing to reconstruct original Prolog terms.

**Neutral**

None

### Dual format: A Hybrid approach with Query-Selectable options

This option combines the aforementioned formats, offering a choice based on request parameters.

Example:

```json
[
  {
    "substitutions": [
      {
        "variable": "X",
        "expression": "1"
        "term": [
          {
            "name": "1"
          }
        ]
      }
    ]
  }
]
```

#### Consequences

**Positive**

- Provides versatility, allowing clients or users to select the most appropriate format for their needs.

**Negative**

More complex in terms of implementation and maintenance.

**Neutral**

None

### Request for community feedback

We seek input on the following:

- Experiences or concerns regarding each response format.
- Alignment of these formats with your usage expectations or needs.
- Requirements for a response format from a developer's perspective.

Any additional suggestions or ideas for alternative response formats or enhancements.

### Future Actions 🚀

We invite all community members to contribute their insights, experiences, and recommendations regarding this topic. Your participation is invaluable in shaping the evolution of the Logic module's response formats, ensuring they meet the diverse needs and expectations of everyone.

**Comment by ccamel** [comment]
@amimart, we should progress on this topic to provide a stable solution. I'm leaning towards using unstructured responses (string-based) because it's simpler and more natural, and I believe implementing a parser wouldn't be too difficult. And as we'll be providing SDKs, with Rust for the contract side and Go for a more general SDK, that should be fine. What do you think?

**Reply by amimart** [reply]
I think both structured and unstructured could be suitable as they define a contract. I tend to prefer the unstructured way for the reasons you mentioned before:
- Reduced payloads;
- DX => responses are human readable;

And regarding interoperability, once we provide a parser it'd be even easier to integrate with than structured reponses. BTW we already have the parser in Rust, won't be difficult to make one in Go :)

**Comment by ccamel** [comment]
Implemented in https://github.com/okp4/okp4d/pull/555.
